### DISCUSSION TITLE
Does the export of the energy dashboard allow for import with your script?

### DISCUSSION THREAD

**Discussion by riemers** [patrickvorgers/Home-Assistant-Import-Energy-Data#55] [General]
As the topic states. I have an older backup which is good but i want the history from that one, as i restore it then it will work a bit then all my data is gone. Must be some sqlite issue i had during a recovery or something. So i have the data but it was still with an old disabled integration (which is why i never removed it) so now i have my new data just a week, but nothing more. I want to make a backup of the current week, restore the old one with years of data. Then export the csv from the energy dashboard and then import it with this script in my new setup. Does that work?

**Comment by patrickvorgers** [comment]
That should be possible but using a slightly different approach. You keep the current database (week of data) and export the data from the old database and import that data into the new database.

You need to follow the "Home Assistant datasource" howto to export the existing data from the "old" SQLite database into the correct CSV files. After that you can follow the generic howto to import those generated CSV files. You have to make sure that you have at least one week of data in the new database because of the shortterm statistics.

**Comment by riemers** [comment]
So long story short, i cannot use the export button that generates the csv file from the energy dashboard? (i am fine with using the instructions thhough, but might be nice to mention that in there too)

**Reply by riemers** [reply]
nevermind, i just looked at the csv generated and there just a snapshot of a period. So no data points. I will do what you requested to fix my 'issue' :)

**Reply by patrickvorgers** [reply]
Sorry, I misunderstood your question.

The problem with the energy dashboard CSV is that it only exports the data in view. You have to export a lot of timeframes in case you want to have a detailled export. For instance: export every day, if you want to have the hourly statistics.

To keep your current data and resolution it is best to use the scripts.

**Reply by riemers** [reply]
I know there is an api endpoint for exports of the data in home assistant, but that might be too far fetched in this case. Because then you would have to check where there is data and not etc etc.. i'll grab the db from an old backup and will just use that one. I can finally get rid of the old integration then too which was on disabled otherwise i would loose the data. I don't know if that has changed by now that they use a virtual energy device as removing say enpase integration or p1 integrations it would remove the data too.

**Reply by patrickvorgers** [reply]
I did some more investigation on the export features of Home Assistant.
You can do the following:

- Go to the "History" tab
- Select "Choose entity" and select the Energy Dashboard entity you want to export
- Adjust the "Start date" and "End date"
- In the top-right corner select the 3-dots
- Select "Download data"

This will export the data in the history view in a CSV file. This CSV file can be easily imported into Home Assistant with some configuration in the generic script. I will have a look whether I can add that script to the "Home Assistant datasource"

**Reply by patrickvorgers** [reply]
I added the needed script and updated the documentation accordingly.
Please let me know if this works for you.

**Reply by riemers** [reply]
Thats a quick one Patrick :) Thanks, i'll try it out

**Reply by riemers** [reply]
Hmm i think there is more going on, with my side it seems: 
![image](https://github.com/user-attachments/assets/a46f401f-9819-4f8b-b8c3-da2148a6cdcb)

If i go to the energy board and select 2024 for a whole year, it doesn't work. If i go to weeks, it gets stuck in september. I can continue to  go in week to 2 - 8 september, but the next 2 weeks are exactly the same (which is wrong) and it only drops to nothing in the 3rth week after.

If i select my outputs without september:
![image](https://github.com/user-attachments/assets/2264be28-1c3d-4cb1-8ef5-c352b63be9aa)

(There is a gap in april but that was my bad)

So something is bad in the data itself not allowing me to get everything. Also the entity that does no longer exist but does hold data i cannot export the history from it since it doesn't exists anymore.

**Reply by patrickvorgers** [reply]
Could it be that your old database is corrupted?

What you could try to do is make multiple exports and exclude the date(s) that cause any trouble. With the script I provided you can generate the import CSV file based on multiple export files.

So basically you have history1.csv history2.csv etc.

With the datasource script you can do: `python HomeAssistDataPrepare.py history*.csv`

I presume that the history export of the 1-1-2021 till 31-08-2024 works, correct?

**Comment by riemers** [comment]
Till August should work yes, but maybe it’s better I take it from the SQLite database? Then I can see what wrong data it might have…?

Verzonden vanuit Outlook voor iOS<https://aka.ms/o0ukef>
________________________________
Van: Patrick Vorgers ***@***.***>
Verzonden: Wednesday, October 2, 2024 6:13:24 PM
Aan: patrickvorgers/Home-Assistant-Import-Energy-Data ***@***.***>
CC: Erik-jan Riemers ***@***.***>; State change ***@***.***>
Onderwerp: Re: [patrickvorgers/Home-Assistant-Import-Energy-Data] Does the export of the energy dashboard allow for import with your script? (Discussion #55)


Could it be that your old database is corrupted?

What you could try to do is make multiple exports and exclude the date(s) that cause any trouble. With the script I provided you can generate the import CSV file based on multiple export files.

So basically you have history1.csv history2.csv etc.

With the datasource script you can do: python HomeAssistDataPrepare.py history*.csv

I presume that the history export of the 1-1-2021 till 31-08-2024 works, correct?

**Reply by patrickvorgers** [reply]
You probably won't be able to access that data in case the database is corrupt but you can always try. 🤞

**Reply by riemers** [reply]
I can see the data, 5.6mil records, but when i say do a count for example it fails. Any query to give me the list of an id results in 0 results. I think at some point homeassistant agrees that its foobar and starts over. I checked my current setup for the db file which has lost it and i see this:
![image](https://github.com/user-attachments/assets/a41c6885-df26-4dc8-991e-053704360c3a)

9-23 it become corrupt for whatever reason. at 02:00 at night..  i'll see if i have an older backup too.

using `sqlite3 db-backup.db ".recover" |sqlite3 new.db` i did manage to get a 1gb file back, trying to see if i can make some use out of it.

**Reply by patrickvorgers** [reply]
I had the same issue a few years back and almost lost all of my data. SQLite is prone to errors when the database is not closed properly. Hopefully you can recover all the data. My advice would be to switch to MariaDB, I never had an issue since.
